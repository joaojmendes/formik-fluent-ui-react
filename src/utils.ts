import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
