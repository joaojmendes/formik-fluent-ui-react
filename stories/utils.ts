import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
