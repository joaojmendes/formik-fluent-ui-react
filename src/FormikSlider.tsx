import { ISliderProps, Slider } from  '@fluentui/react';
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, Omit } from './utils'

export function mapFieldToSlider<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<ISliderProps, 'value' | 'onChange' | 'onChanged'> {
  return {
    value: field.value,
    onChange: value => {
      form.setFieldValue(field.name, value)
    },
    onChanged: () => field.onBlur(createFakeEvent(field)),
  }
}

export type FormikSliderProps<T = any> = Omit<
  ISliderProps,
  'value' | 'onChange' | 'onChanged'
> &
  FieldProps<T>

export function FormikSlider<T = any>({
  field,
  form,
  ...props
}: FormikSliderProps<T>) {
  return <Slider {...props} {...mapFieldToSlider({ field, form })} />
}
