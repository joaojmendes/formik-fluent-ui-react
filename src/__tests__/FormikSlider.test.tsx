import { Field, FieldProps, Form, Formik } from 'formik'
import { Slider } from 'office-ui-fabric-react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikSlider, mapFieldToSlider } from '../FormikSlider'
import { noop, serialize } from './utils'

class Values {
  public rating: number = 3
}

function createFieldProps(value: number = 3): FieldProps<{ test: Date }> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'rating',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

test('<FormikSlider /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field name="rating" label="Rating" component={FormikSlider} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikSlider /> renders a Fabric <Slider />', () => {
  const fieldProps = createFieldProps()

  const formikSlider = renderer.create(
    <FormikSlider {...fieldProps} label="Rating" />
  )
  const fabricSlider = renderer.create(
    <Slider {...mapFieldToSlider(fieldProps)} label="Rating" />
  )
  expect(serialize(formikSlider)).toBe(serialize(fabricSlider))
})

test('mapFieldToSlider() maps FieldProps to ISliderProps', () => {
  const { field, form } = createFieldProps()
  const props = mapFieldToSlider({ form, field })

  expect(props.value).toBe(field.value)

  props.onChange!(10)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, 10)

  props.onChanged!(null as any, 10)
  // onChanged calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
