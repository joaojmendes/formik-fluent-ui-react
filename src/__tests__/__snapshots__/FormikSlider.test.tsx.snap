// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikSlider /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-Slider ms-Slider-enabled ms-Slider-row root-47"
  >
    <label
      className="ms-Label titleLabel-57"
      htmlFor="Slider1"
    >
      Rating
    </label>
    <div
      className="ms-Slider-container container-49"
    >
      <div
        aria-disabled={false}
        aria-label="Rating"
        aria-valuemax={10}
        aria-valuemin={0}
        aria-valuenow={3}
        className="ms-Slider-slideBox ms-Slider-showValue ms-Slider-showTransitions slideBox-50"
        data-is-focusable={true}
        id="Slider1"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
      >
        <div
          className="ms-Slider-line line-52"
        >
          <span
            className="ms-Slider-thumb thumb-51"
            style={
              Object {
                "left": "30%",
              }
            }
          />
          <span
            className="lineContainer-53 ms-Slider-active activeSection-54"
            style={
              Object {
                "width": "30%",
              }
            }
          />
          <span
            className="lineContainer-53 ms-Slider-inactive inactiveSection-55"
            style={
              Object {
                "width": "70%",
              }
            }
          />
        </div>
      </div>
      <label
        className="ms-Label ms-Slider-value valueLabel-58"
      >
        3
      </label>
    </div>
  </div>
</form>
`;
